src/ArrowFunctions.cpp

```diff
@@ -275,25 +275,47 @@ int64_t cpp_getStringColumnNumBytes(const char* filename, const char* colname, v
         }
         column_reader = row_group_reader->Column(idx);
 
-        int16_t definition_level;
         parquet::ByteArrayReader* ba_reader =
           static_cast<parquet::ByteArrayReader*>(column_reader.get());
 
         int64_t numRead = 0;
         while (ba_reader->HasNext() && numRead < numElems) {
-          parquet::ByteArray value;
-          (void)ba_reader->ReadBatch(1, &definition_level, nullptr, &value, &values_read);
-          if ((ty == ARROWLIST && definition_level == 3) || ty == ARROWSTRING) {
-            if(values_read > 0) {
-              offsets[i] = value.len + 1;
-              byteSize += value.len + 1;
-              numRead += values_read;
-            } else {
-              offsets[i] = 1;
-              byteSize+=1;
-              numRead+=1;
+          if((numElems - numRead) < batchSize)
+            batchSize = numElems-numRead;
+          std::vector<parquet::ByteArray> string_values(batchSize);
+          std::vector<int16_t> definition_level(batchSize);
+          (void)ba_reader->ReadBatch(batchSize, definition_level.data(), nullptr, string_values.data(), &values_read);
+          numRead += values_read;
+          if (ty == ARROWSTRING) {
+            auto numCols = file_metadata -> num_columns();
+            int string_index = 0;
+            for(int idx = 0; idx < definition_level.size(); idx++) {
```

Comment:
This doens't need to be updated because they should always be equivalent but I feel like `idx < batchSize` is 
a bit clearer. But since we're using `idx` to indexing to `definition_level` maybe this is better? idk i don't have terribly strong feeligns

Reply:
Well, the thing is that `idx < batchSize` is not the same as `idx < definition_level.size()`. When you read a batch in Parquet, sometimes some of the reads will "fail", so you don't actually read a full `batchSize` in.

The `definition_level.size()` will return the number of elements that were actually read in and initialized by the `ReadBatch` call, so you can think of something like `ReadBatch(10,...)` actually only reading in 6 values or something like that, which would result in `definition_level.size()` being 6, so that is why we have that distinction here.

Reply:
oh gotcha! Yah def leave it as that then 😁